File: Hashing/php/intersectionWithDupes.php
<?php

declare(strict_types=1);

function intersectionWithDupes(array $first, array $second): array
{
    $counts = [];
    foreach ($first as $item) {
        $key = serialize($item);
        $counts[$key] = ($counts[$key] ?? 0) + 1;
    }

    $result = [];
    foreach ($second as $item) {
        $key = serialize($item);
        if (($counts[$key] ?? 0) > 0) {
            $result[] = $item;
            $counts[$key]--;
        }
    }

    return $result;
}

File: Hashing/csharp/IntersectionWithDupes.cs
using System.Collections.Generic;

namespace HashingSolutions;

public static class IntersectionWithDupes
{
    public static IList<T> Solve<T>(IEnumerable<T> first, IEnumerable<T> second)
    {
        var counts = new Dictionary<T, int>();
        foreach (var value in first)
        {
            counts.TryGetValue(value, out int current);
            counts[value] = current + 1;
        }

        var result = new List<T>();
        foreach (var value in second)
        {
            if (counts.TryGetValue(value, out int remaining) && remaining > 0)
            {
                result.Add(value);
                counts[value] = remaining - 1;
            }
        }

        return result;
    }
}
